ethereum/contracts/Broker.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.6.2;
+
+import "./Verifier.sol";
+import "./Decoder.sol";
+import "./Application.sol";
+
+contract Broker {
+    using Decoder for bytes;
+
+    mapping(address => bool) public applications;
+    Verifier public verifier;
+
+    constructor(address verifierAddr) public {
+        verifier = new Verifier(verifierAddr);
+    }
+
+    /**
+     * @dev registers new applications
+     * @param _appID address _appID is the contract address to be registered as a supported application
+     */
+    function registerApp(address _appID)
+        public
+    {
+        require(!applications[_appID], "Application is already registered");
+
+        // 0x0 is the value returned for not-yet created accounts.
+        // 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470 is returned for accounts without code.
+        bytes32 accountHash = 0xc5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470;
+        bytes32 codehash;
+        assembly { codehash := extcodehash(_appID) }
+        require(codehash != 0x0 && codehash != accountHash, "Application must be a contract deployed to this network");
+
+        // TODO: Consider checking that the application implements the abstract Application method 'submit' according
+        //       to EIP-165 (incomplete implementation: https://github.com/ethereum/EIPs/blob/master/EIPS/eip-165.md)
+
+        applications[_appID] = true;
+    }
+
+    /**
+     * @dev routes the message to the specified application ID after verifying the operator's signature
+     * @param _data address _data expected type: Message { AppID [32]byte, Payload []byte }
+     * @param _signature address _signature expected type: Signature []byte
+     */
+    function submit(bytes memory _data, bytes memory _signature)
```

Comment:
Looks like relayer will call this method to submit messages to the Ethereum side.

On the Substrate side, things are bit different. We have a Bridge module that accepts messages from the Relayer, and then forwards those messages to the Broker module. If we want to follow that, then I'd suggest the following structure:

Bridge Contract:
```
function send(address _app_id, bytes memory _message) {
  emit MessageReceived(keccak_hash(_message))
  broker.submit(address _app_id, bytes memory _message)
}
```

Broker Contract:
```
function submit(address _app_id, bytes memory _message) {
  // dispatch to target app
}
```

But whether we want a Bridge contract is up for debate. Though it would be nice to mirror the substrate side. @musnit any thoughts from your side?

Reply:
i think for simplicity we should keep it mirrored for now. if we want to modify in future for other reasons, can do so, but would prefer for it to be symmetrical and clean at least initially if there are no side effects to that.

Reply:
This feels like an arbitrary waste of gas.  Calling a function on another contract and passing it the payload increases total gas cost. Why not just emit the event on the Broker contract or include `function send` on the Broker contract?

As gas isn't a priority right now I'm happy to add the Bridge contract to mirror the substrate side, but we should be aware of the implicit consequences.

Reply:
would it be worth considering the inverse, ie, is it up for debate whether we even want a Bridge **pallet**?

Reply:
Come to think of it, all the bridge pallet really does is forward the message to the broker. So yeah I wouldn't mind removing it. Though we should then rename the Broker contract to Bridge.

I'm up for it.

Reply:
Currently my the latest parachain refactor has the following message flow:
```
Bridge:
  send(msg):
    broker.submit(msg)
    
Broker:
  submit(msg):
    if verifier.verify(msg) == true
      self.dispatch_to_app(msg)
```

If we ditch the old Bridge contract, and rename things, then the flow would look like this:
```
Bridge:
  submit(msg):
    if verifier.verify(msg)
      self.dispatch_to_app(msg)
```

Reply:
In anticipation of this change on the Substrate side I've renamed Broker.sol to Bridge.sol in https://github.com/Snowfork/polkadot-ethereum/pull/75/commits/f5c731cfa11dfda704bca49a681da2e857c20d64

Reply:
Same here. I've done the same on the Substrate side.